invoice/invoice.go

```
@@ -0,0 +1,966 @@
+				// supported one.
+				continue
+			}
+
+			base256Data, err := bech32.ConvertBits(base32Data, 5, 8,
+				false)
+			if err != nil {
+				return err
+			}
+
+			for len(base256Data) > 0 {
+				info := ExtraRoutingInfo{}
+				info.PubKey, err = btcec.ParsePubKey(
+					base256Data[:33], btcec.S256())
+				if err != nil {
+					return err
+				}
+				info.ShortChanID = binary.BigEndian.Uint64(
+					base256Data[33:41])
+				info.Fee = binary.BigEndian.Uint64(
+					base256Data[41:49])
+				info.CltvExpDelta = binary.BigEndian.Uint16(
+					base256Data[49:51])
+				invoice.RoutingInfo = append(
+					invoice.RoutingInfo, info)
+				base256Data = base256Data[51:]
+			}
+		default:
+			// Ignore unknown type.
+		}
+	}
+
+	return nil
+}
+
+// writeTaggedFields writes the non-nil tagged fields of the Invoice to the
+// base32 buffer.
+func writeTaggedFields(bufferBase32 *bytes.Buffer, invoice *Invoice) error {
+	if invoice.PaymentHash != nil {
+		// Convert 32 byte hash to 52 5-bit groups.
+		base32, err := bech32.ConvertBits(invoice.PaymentHash, 8, 5,
+			true)
+		if err != nil {
+			return err
+		}
+		if len(base32) != 52 {
+			return fmt.Errorf("invalid payment hash length: %d",
+				len(invoice.PaymentHash))
+		}
+
+		err = writeTaggedField(bufferBase32, fieldTypeP, base32)
+		if err != nil {
+			return err
+		}
+	}
+
+	if invoice.Description != nil {
+		base32, err := bech32.ConvertBits([]byte(*invoice.Description),
+			8, 5, true)
+		if err != nil {
+			return err
+		}
+		writeTaggedField(bufferBase32, fieldTypeD, base32)
+	}
+
+	if invoice.DescriptionHash != nil {
+		// Convert 32 byte hash to 52 5-bit groups.
+		descBase32, err := bech32.ConvertBits(invoice.DescriptionHash,
+			8, 5, true)
+		if err != nil {
+			return err
+		}
+
+		if len(descBase32) != 52 {
+			return fmt.Errorf("invalid description hash length: %d",
+				len(invoice.DescriptionHash))
+		}
+
+		writeTaggedField(bufferBase32, fieldTypeH, descBase32)
```

Ignoring an error here, repeated in a few other instances in this function (won't comment on other instances to avoid comment spam).